### PR TITLE
feat(destination): onboard campaign manager

### PR DESCRIPTION
**Pull request by aashishmalik**
## Description of the change

> Onboard Campaign Manager

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/destinations/campaign_manager/data/DCMTrackConfig.json [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  {
+    "destKey": "mobileDeviceId",
+    "sourceKeys": "properties.mobileDeviceId",
+    "required": false
+  },
+  {
+    "destKey": "value",
+    "sourceKeys": "properties.value",
+    "required": false
+  },
```
[the remark this answers is not in the document]
Comment (aashishmalik) [reply v0/destinations/campaign_manager/data/DCMTrackConfig.json]: I have kept it to be in sync with official documentation.  Are you suggesting pick any one of these 3 . In that case there might be multiple keys and will involve giving preference to 1 property.

File: v0/destinations/campaign_manager/transform.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  response.endpoint = endpointUrl;
+  response.headers = {
+    Authorization: `Bearer ${getAccessToken(metadata)}`,
+    "Content-Type": "application/json"
+  };
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.body.JSON.kind =
+    requestJson.messageType === "batchinsert"
+      ? "dfareporting#conversionsBatchInsertRequest"
+      : "dfareporting#conversionsBatchInsertRequest";
```
[the remark this answers is not in the document]
Comment (aashishmalik) [reply v0/destinations/campaign_manager/transform.js]: properties.requestType will decide between update and requests.

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  },
+  {
+    "destKey": "ordinal",
+    "sourceKeys": "properties.ordinal",
+    "required": true
+  },
+  {
+    "destKey": "timestampMicros",
+    "sourceKeys": "originalTimestamp",
+    "required": true
```
Comment (utsabc) [inline v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json:15]:
```suggestion
    "required": true,
    "metadata": {   "type": "microSecondTimestamp"  }
     
```

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  },
+  {
+    "destKey": "mobileDeviceId",
+    "sourceKeys": "properties.mobileDeviceId",
+    "required": false
+  },
+  {
+    "destKey": "quantity",
+    "sourceKeys": "properties.quantity",
+    "required": false
```
Comment (utsabc) [inline v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json:35]:
```suggestion
    "destKey": "quantity",
    "sourceKeys": "properties.quantity",
    "metadata": {
        "type": "toNumber"
      },
    "required": false
```

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  },
+  {
+    "destKey": "value",
+    "sourceKeys": "properties.value",
+    "required": false
```
Comment (utsabc) [inline v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json:40]: Same here
```suggestion
    "destKey": "value",
   "sourceKeys": [
        "properties.total",
        "properties.value",
        "properties.revenue"
      ],
      "metadata": {
        "type": "toNumber"
      },
    "required": false
```

File: v0/destinations/campaign_manager/transform.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    requestJson.limitAdTracking || destination.limitAdTracking;
+  requestJson.treatmentForUnderage =
+    requestJson.treatmentForUnderage || destination.treatmentForUnderage;
+  requestJson.childDirectedTreatment =
+    requestJson.childDirectedTreatment || destination.childDirectedTreatment;
+  requestJson.nonPersonalizedAd =
+    requestJson.nonPersonalizedAd || destination.nonPersonalizedAd;
+  const endpointUrl = prepareUrl(message);
+  // to unix epoc micro seconds
+  requestJson.timestampMicros = (
```
Comment (utsabc) [inline v0/destinations/campaign_manager/transform.js:67]: This will not be needed

File: v0/destinations/campaign_manager/networkHandler.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    if (
+      Array.isArray(response.status) &&
+      Array.isArray(response.status[0].errors)
+    ) {
+      return (
+        response.status[0].errors[0].code !== "PERMISSION_DENIED" &&
+        response.status[0].errors[0].code !== "INVALID_ARGUMENT"
+      );
+    }
+    return false;
```
Comment (utsabc) [inline v0/destinations/campaign_manager/networkHandler.js:47]: return true

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json (hunk 5 above)
Comment (aashishmalik) [reply v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json]: done

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json (hunk 4 above)
Comment (aashishmalik) [reply v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json:38]: this should be string

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json (hunk 3 above)
Comment (aashishmalik) [reply v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json]: done

**Review by utsabc — DISMISSED** [review]

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    "destKey": "quantity",
+    "sourceKeys": "properties.quantity",
+    "required": false
+  },
+  {
+    "destKey": "value",
+    "sourceKeys": [
+      "properties.total",
+      "properties.value",
+      "properties.revenue"
```
Comment (ItsSudip) [inline v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json:45]:
```suggestion
      "properties.value",
      "properties.total",
      "properties.revenue"
```

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json (hunk 8 above)
Comment (aashishmalik) [reply v0/destinations/campaign_manager/data/CampaignManagerTrackConfig.json]: done

**Review by ItsSudip — APPROVED** [review]

**Review by ItsSudip — DISMISSED** [review]

**Review by utsabc — DISMISSED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**